### PR TITLE
sqlite [WIP] #nomerge

### PR DESCRIPTION
**Pull request by codegod100**

### CODE REVIEW

File: app/db.py [hunk 1]
```diff
@@ -75,57 +84,63 @@ def n_edges(self):
...
         # example use: if [[server-agora]] does not exist, serve [[agora-server]]
         #
         # this is, of course, a terrible implementation and dangerous :)
         # only enable up to 4 tokens as 24 permutations is manageable.
         tokens = uri.split('-')
         permutations = itertools.permutations(tokens, max_length)
         permutations = ['-'.join(permutation) for permutation in permutations]
-        nodes = [node for node in G.nodes(only_canonical=True).values() if node.wikilink in permutations and node.subnodes]
+        nodes = [node for node in G.nodes(only_canonical=True).values(
+        ) if node.wikilink in permutations and node.subnodes]
```
Comment (flancian) [inline app/db.py:109]: This doesn't look nice to me, what's up with the reformatting?

File: app/db.py [hunk 2]
```diff
@@ -75,57 +84,63 @@ def n_edges(self):
...
     def search(self, regex):
         # returns a list of nodes reasonably freely matching a regex.
-        current_app.logger.debug(f'*** Looking for nodes matching {regex} freely.')
-        nodes = [node for node in G.nodes(only_canonical=True).values() if 
-                    # has some content
-                    node.subnodes and
-                    # its wikilink matches the regex
-                    re.search(regex, node.wikilink)
-                ]
+        current_app.logger.debug(
```
Comment (flancian) [inline app/db.py:126]: What's up with these? Is your editor enforcing 80 characters? Can you tell it not to do that? :) It makes the diff harder to read. Reformatting of existing code should ideally be done in a different PR.

File: app/agora.py [hunk 3]
```diff
@@ -76,24 +81,27 @@ def build_node(node, extension='', user_list=''):
     node = urllib.parse.unquote_plus(node)
     node = util.slugify(node)
 
-    n = copy(G.node(node))
+    # n = copy(G.node(node))
+    n = db.query_node(node)
```
Comment (flancian) [inline app/agora.py:85]: IIUC this PR adds a hard dependency on the .sqlite file for the Agora to work. I don't think we want to do that. The Agora should continue to work (albeit slowly/with a reduced feature set) without .sqlite -- at least while we're experimenting. Then we can have a checkpoint to make the .sqlite file a hard dependency. Until then, I think we should use sqlite where available but keep code paths free of sqlite and working. Ideally we should also have an ENABLE_SQLITE in config.py to make this easier to manage. Wdyt?

**Review by flancian — COMMENTED** [review]
Thank you for starting on this! I left some comments, the main one being in agora.py. Wdyt?

File: app/agora.py (hunk 3 above)
Comment (fuck-capitalism) [reply app/agora.py:85]: I keep vacillating on this. if its optional I don't think we should even add sqlite. just needless complexity without measurable increase.

File: app/db.py (hunk 1 above)
Comment (codegod100) [reply app/db.py:109]: probably auto formatter I think it uses pep8?

Comment (codegod100) [reply app/db.py:109]: this why I like that gofmt is programmatically enforced. it parses the AST not just line formatting like most language formatters

File: app/db.py (hunk 2 above)
Comment (codegod100) [reply app/db.py:126]: this isn't on purpose. blame vscode

File: app/agora.py (hunk 3 above)
Comment (codegod100) [reply app/agora.py:85]: I keep vacillating on this. if its optional I don't think we should even add sqlite. just needless complexity without measurable increase.

Comment (flancian) [reply app/agora.py:85]: I know what you mean! But I keep wondering: why can't it be like Javascript? The Agora runs without Javascript; it's just better (more featureful) when you enable it. We could take the same approach with sqlite/other databases? The Agora works on the filesystem only, which makes it so that starting an Agora is just launching the server and making sure the right files are in the right paths; but if you run an extra process that generates sqlite, you get goodies.

File: app/db.py (hunk 2 above)
Comment (flancian) [reply app/db.py:126]: 👍

File: app/agora.py (hunk 3 above)
Comment (codegod100) [reply app/agora.py:85]: >I know what you mean! But I keep wondering: why can't it be like Javascript? The Agora runs without Javascript; The Agora runs without Javascript; it's just better (more featureful) when you enable it.

I kinda feel like this is saying "computers run without an operating system, they're just more featureful when you have one"
